## Project Description
The company bought a large subscription book reading service. 
- It is necessary to analyze a database that contains information about books, publishers, authors, as well as user reviews of books.

### The purpose of the study:

The aim of the project is to analyze a database containing information about books, publishers, authors and user reviews of books.

## Data review

In [1]:

import pandas as pd
from sqlalchemy import create_engine

In [2]:
# set parameters
db_config = {'user': 'praktikum_student', # username
'pwd': 'Sdf4$2;d-d30pp', # password
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # connection port
'db': 'data-analyst-final-project-db'} # database name
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

In [3]:

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
def read_sql(query):
    result = pd.io.sql.read_sql(query, con = engine)
    return result

Let's look at the database tables and their contents

In [5]:
books = """
SELECT *
FROM books
LIMIT 5;
"""
read_sql(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
query = """
SELECT COUNT(book_id) AS book_count
FROM books;
"""
result = read_sql(query)
book_count = result.iloc[0, 0] # get the value from the first row and first column
print(f"Number of books: {book_count}")

Number of books: 1000


In [7]:
query = """
SELECT COUNT(DISTINCT author_id) AS author_count
FROM books;
"""
result = read_sql(query)
author_count = result.iloc[0, 0] # get the value from the first row and first column
print(f"Number of authors: {author_count}")

Number of authors: 636


In [8]:
query = """
SELECT COUNT(DISTINCT publisher_id) AS publisher_count
FROM books;
"""
result = read_sql(query)
publishers_count = result.iloc[0, 0] # get the value from the first row and first column
print(f"Number of publishers: {publishers_count}")

Number of publishers: 340


In [9]:
query = """
SELECT MIN(publication_date) AS min_date, MAX(publication_date) AS max_date
FROM books;
"""
result = read_sql(query)
min_date = result.iloc[0, 0] # get the minimum date value from the first row and first column
max_date = result.iloc[0, 1] # get the maximum date value from the first row and second column
print(f"Minimum publication date: {min_date}")
print(f"Maximum publication date: {max_date}")

Minimum publication date: 1952-12-01
Maximum publication date: 2020-03-31


The `books` table

Contains data about books:

- `book_id' — book id;
- `author_id` — the author's ID;
- `title` — the title of the book;
- `num_pages` — number of pages;
- `publication_date' — publication date of the book;
- `publisher_id' is the ID of the publisher.

The table shows
- 1000 books
- 636 authors
- - 340 publishing house
- Publication dates - from December 1, 1952 to March 31, 2020.

In [10]:
authors = """
SELECT *
FROM authors
LIMIT 5;
"""
read_sql(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [11]:
query = """
SELECT COUNT (*) AS length
FROM authors;
"""
read_sql(query) 

length
0     636

The `authors` table

Contains information about the authors:

- `author_id` — the author's ID;
- `author' — the author's name.

In [12]:
publishers = """
SELECT *
FROM publishers
LIMIT 5;
"""
read_sql(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

The `publishers` table

Contains data about publishing houses:

- `publisher_id' — publisher ID;
- `publisher' — the name of the publisher;

In [13]:
ratings = """
SELECT *
FROM ratings
LIMIT 5;
"""
read_sql(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [14]:
query = """
SELECT MIN(rating) AS min_rating, MAX(rating) AS max_rating
FROM ratings;
"""
result = read_sql(query)
min_rating = result.iloc[0, 0] # get the minimum date value from the first row and first column
max_rating = result.iloc[0, 1] # get the maximum date value from the first row and second column
print(f"Minimum rating: {min_rating}")
print(f"Maximum rating: {max_rating}")

Minimum rating: 1
Maximum rating: 5


In [15]:
query = """
SELECT COUNT(rating) AS rating_count
FROM ratings;
"""
result = read_sql(query)
rating_count = result.iloc[0, 0] # get the value from the first row and first column
print(f"Number of ratings given: {rating_count}")

Number of ratings given: 6456


The `ratings` table

Contains data about user ratings of books:

- `rating_id' — rating ID;
- `book_id'— book id;
- `username` — the name of the user who left the rating;
- `rating' — rating of the book.
- Minimum rating of 1
- Maximum 5
- - Number of users rating: 160
- The number of their ratings: 6456

In [16]:
reviews = """
SELECT *
FROM reviews
LIMIT 5;
"""
read_sql(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [17]:
query = """
SELECT COUNT(DISTINCT username) AS users_count
FROM reviews;
"""
result = read_sql(query)
user_count = result.iloc[0, 0] # get the value from the first row and first column
print(f"Number of users writing reviews: {user_count}")

Number of users writing reviews: 160


In [18]:
query = """
SELECT COUNT(text) AS text_count
FROM reviews;
"""
result = read_sql(query)
text_count = result.iloc[0, 0] # get the value from the first row and first column
print(f"Number of reviews written: {text_count}")

Number of reviews written: 2793


The `reviews` table

Contains data about user reviews of books:

- `review_id' — review ID;
- `book_id' — book id;
- `username` — the name of the user who wrote the review;
- `text' — the text of the review.
- Number of reviews written: 2793

## Conclusion:
- The database contains 5 tables: "books", "authors", "ratings", "publishers" and "reviews".
- The "books" table contains information about 1000 books.
- The table "authors" contains data on 636 authors.
- The table "publishers" contains information about 340 publishers.
- The range of publication dates of books is from December 1, 1952 to March 31, 2020.
- The number of unique users who have evaluated and written book reviews is 160.
- The number of ratings: 6456
- Book scores range from 1 to 5 points.
- Written reviews 2793

### Counting how many books were published after January 1, 2000

In [19]:
query = """
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'::date
"""
read_sql(query)

count
0    819

### Conclusion:
- The number of books published after January 1, 2000 is 819.

### For each book, we will calculate the number of reviews and the average rating

In [20]:
query = """
SELECT b.title,
       COUNT(DISTINCT rv.review_id) AS review_cnt,
       ROUND(AVG(r.rating), 2) AS rating_avg
FROM books AS b
LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
LEFT JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY b.book_id
ORDER BY rating_avg DESC, review_cnt DESC
"""
read_sql(query)


title  review_cnt  rating_avg
0                        A Dirty Job (Grim Reaper  #1)           4        5.00
1              School's Out—Forever (Maximum Ride  #2)           3        5.00
2         Moneyball: The Art of Winning an Unfair Game           3        5.00
3                   The Big Bad Wolf  (Alex Cross  #9)           2        5.00
4                 Welcome to Temptation (Dempseys  #1)           2        5.00
..                                                 ...         ...         ...
995  The World Is Flat: A Brief History of the Twen...           3        2.25
996                                      Drowning Ruth           3        2.00
997                  His Excellency: George Washington           2        2.00
998                                              Junky           2        2.00
999                               Harvesting the Heart           2        1.50

[1000 rows x 3 columns]

### Conclusion:
We have access to 1000 books, which are sorted by average rating and number of reviews. The book "A Dirty Job (Grim Reaper #1)" has the highest rating and the largest number of reviews among books with an average rating of 5.

### Determine the publishing house that has released the largest number of books. 
- At the same time, we will exclude brochures from the analysis and consider only books with more than 50 pages.

In [21]:
query = """
SELECT publisher, COUNT(book_id) AS books_cnt
FROM publishers AS p
INNER JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY books_cnt DESC
LIMIT 1
"""
read_sql(query)

publisher  books_cnt
0  Penguin Books         42

### Conclusion:
The publishing house that has released the largest number of books is Penguin Books (42 books)

### Determine the author who has the highest average score for books with 50 or more ratings.

In [22]:
query = """
SELECT a.author, AVG(r.rating) AS mean_rating
FROM authors AS a
JOIN books AS b ON a.author_id = b.author_id
JOIN ratings AS r ON b.book_id = r.book_id
WHERE b.book_id IN (
    SELECT book_id
    FROM ratings 
    GROUP BY book_id
    HAVING COUNT(rating_id) >= 50
)
GROUP BY a.author
ORDER BY mean_rating DESC
LIMIT 1
"""
read_sql(query)

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.287097

### Output:
The author who has the highest average score for books with 50 or more ratings, J.K. Rowling/Mary GrandPré average score(4.287097)

### Calculate the average number of reviews from active users. 
- We consider those who have left more than 48 ratings to be active users.

In [23]:
query = """
SELECT ROUND(AVG(review_cnt))
FROM (
    SELECT COUNT(review_id) AS review_cnt
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 48
    )
    GROUP BY username
) AS review_count
"""
read_sql(query)



round
0   24.0

### Conclusion:
Among active users who have left more than 48 ratings, on average, everyone writes 24 reviews.

## General conclusion:
Examining the database containing information about books, authors, publishers, ratings and reviews, we obtained the following results:

- The database contains 5 tables: "books", "authors", "ratings", "publishers" and "reviews".
- The "books" table contains information about 1000 books.
- The "authors" table contains data on 636 authors.
- The table "publishing houses" contains information about 340 publishing houses.
- The range of publication dates of books covers the period from December 1, 1952 to March 31, 2020.
- The number of unique users who have rated books and written reviews is 160.
- The total number of ratings set by users is 6456.
- Book ratings range from 1 to 5 points.
- There are 2793 written book reviews in the database.

Based on the analysis, we can draw the following conclusions:

1. The number of books released after January 1, 2000 is 819.
2. Among the available books, the book "A Dirty Job (Grim Reaper #1)" has the highest rating and the largest number of reviews among books with a rating of 5.
3. The publishing house "Penguin Books" has published the largest number of books (42 books).
4. Author J.K. Rowling/Mary GrandPré has the highest average rating for books with 50 or more ratings (average rating 4.29).
5. Among active users who have left more than 48 ratings, on average each user writes 24 reviews.